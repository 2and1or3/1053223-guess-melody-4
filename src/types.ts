enum GameType {
  ARTIST = `artist`,
  GENRE = `genre`,
}

interface GenreAnswer {
  genre: string;
  src: string;
}

interface ArtistAnswer {
  artist: string;
  pictureSrc: string;
}

interface GenreQuestion {
  type: GameType.GENRE;
  genre: string;
  answers: Array<GenreAnswer>;
}

interface ArtistQuestion {
  type: GameType.ARTIST;
  artist: string;
  trackSrc: string;
  answers: Array<ArtistAnswer>;
}

export { GenreAnswer, ArtistAnswer, GenreQuestion, ArtistQuestion, GameType };
