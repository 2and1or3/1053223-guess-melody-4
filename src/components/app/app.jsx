import React from "react";
import PropTypes from "prop-types";
import {BrowserRouter, Route, Switch} from "react-router-dom";
import {PureComponent} from "react";
import {connect} from "react-redux";

import WelcomeScreen from '../welcome-screen/welcome-screen.jsx';
import QuestionGenreScreen from '../question-genre-screen/question-genre-screen.jsx';
import QuestionArtistScreen from '../question-artist-screen/question-artist-screen.jsx';
import GameScreen from '../game-screen/game-screen.jsx';
import GameOverScreen from '../game-over-screen/game-over-screen.jsx';
import WinScreen from '../win-screen/win-screen.jsx';

import withActivePlayer from '../../hocs/with-active-player/with-active-player.js';
import withUserAnswer from '../../hocs/with-user-answer/with-user-answer.js';
import {ActionCreator} from '../../reducer/game/game.js';
import {getQuestions} from '../../reducer/data/selectors.js';
import {getMistakes, getMaxMistakes, getStep} from '../../reducer/game/selectors.js';

import {GameType} from '../../consts.js';
import {genreProp, artistProp} from '../../props.js';

const QuestionGenreScreenWrapped = withActivePlayer(withUserAnswer(QuestionGenreScreen));
const QuestionArtistScreenWrapped = withActivePlayer(QuestionArtistScreen);


class App extends PureComponent {
  _renderGameScreen() {
    const {questions, maxMistakes, step, onPlayClick, onAnswer, mistakes, onRepeat} = this.props;
    const question = questions[step];

    if (step === -1) {
      return (
        <WelcomeScreen
          errorCount = {maxMistakes}
          onPlayClick = {onPlayClick}
        />
      );
    }

    if (mistakes >= maxMistakes) {
      return (
        <GameOverScreen onRepeat = {onRepeat}/>
      );
    }

    if (step >= questions.length) {
      return (
        <WinScreen onRepeat = {onRepeat} quantity = {questions.length} mistakes = {mistakes}/>
      );
    }

    if (question) {
      switch (question.type) {
        case GameType.ARTIST:
          return (
            <GameScreen gameType = {GameType.ARTIST} maxMistakes = {maxMistakes} mistakes = {mistakes}>
              <QuestionArtistScreenWrapped
                question = {question}
                onAnswer = {onAnswer}
              />
            </GameScreen>
          );
        case GameType.GENRE:
          return (
            <GameScreen gameType = {GameType.GENRE} maxMistakes = {maxMistakes} mistakes = {mistakes}>
              <QuestionGenreScreenWrapped
                question = {question}
                onAnswer = {onAnswer}
              />
            </GameScreen>
          );
      }
    }

    return null;
  }

  render() {
    const {questions} = this.props;

    return (
      <BrowserRouter>
        <Switch>
          <Route exact path="/">
            {this._renderGameScreen()}
          </Route>
          <Route exact path="/artist">
            <GameScreen gameType = {GameType.ARTIST} maxMistakes = {3} mistakes = {0}>
              <QuestionArtistScreenWrapped question = {questions[1]} onAnswer = {() => {}}/>
            </GameScreen>
          </Route>
          <Route exact path="/genre">
            <GameScreen gameType = {GameType.GENRE} maxMistakes = {3} mistakes = {0}>
              <QuestionGenreScreenWrapped question = {questions[0]} onAnswer = {() => {}}/>
            </GameScreen>
          </Route>
        </Switch>
      </BrowserRouter>
    );
  }
}

App.propTypes = {
  mistakes: PropTypes.number.isRequired,
  maxMistakes: PropTypes.number.isRequired,
  step: PropTypes.number.isRequired,
  questions: PropTypes.arrayOf(PropTypes.oneOfType([genreProp, artistProp])).isRequired,
  onPlayClick: PropTypes.func.isRequired,
  onAnswer: PropTypes.func.isRequired,
  onRepeat: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  mistakes: getMistakes(state),
  maxMistakes: getMaxMistakes(state),
  questions: getQuestions(state),
  step: getStep(state),
});

const mapDispatchToProps = (dispatch) => ({
  onPlayClick: () => {
    dispatch(ActionCreator.incrementStep());
  },
  onAnswer: (question, answer) => {
    dispatch(ActionCreator.incrementMistakes(question, answer));
    dispatch(ActionCreator.incrementStep());
  },
  onRepeat: () => {
    dispatch(ActionCreator.repeatGame());
  }
});

const ConnectedApp = connect(mapStateToProps, mapDispatchToProps)(App);

export {App};
export default ConnectedApp;
