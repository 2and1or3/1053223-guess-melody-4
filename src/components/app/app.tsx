import * as React from "react";
import {BrowserRouter, Route, Switch, Redirect} from "react-router-dom";
import {connect} from "react-redux";

import WelcomeScreen from '../welcome-screen/welcome-screen';
import QuestionGenreScreen from '../question-genre-screen/question-genre-screen';
import QuestionArtistScreen from '../question-artist-screen/question-artist-screen';
import GameScreen from '../game-screen/game-screen';
import GameOverScreen from '../game-over-screen/game-over-screen';
import WinScreen from '../win-screen/win-screen';
import AuthorizationScreen from '../authorization-screen/authorization-screen';
import PrivateRoute from '../private-route/private-route';

import withActivePlayer from '../../hocs/with-active-player/with-active-player';
import withUserAnswer from '../../hocs/with-user-answer/with-user-answer';
import {ActionCreator} from '../../reducer/game/game';
import {Operation as UserOperation} from '../../reducer/user/user';
import {getQuestions} from '../../reducer/data/selectors';
import {getMistakes, getMaxMistakes, getStep} from '../../reducer/game/selectors';
import {getUserStatus} from '../../reducer/user/selectors';

import {AuthorizationStatus, AppRoute} from '../../consts';
import {GameType, GenreQuestion, ArtistQuestion} from '../../types';

const QuestionGenreScreenWrapped = withActivePlayer(withUserAnswer(QuestionGenreScreen));
const QuestionArtistScreenWrapped = withActivePlayer(QuestionArtistScreen);

type Question = GenreQuestion | ArtistQuestion;

interface Props {
  mistakes: number;
  maxMistakes: number;
  step: number;
  questions: Question[];
  onPlayClick: () => void;
  onAnswer: (question: Question, answer: string | boolean[]) => void;
  onRepeat: () => void;
  onAuthSubmit: (login: string, password: string) => void;
  onGoToWelcome: () => void;
  userStatus: string;
}

class App extends React.PureComponent<Props> {
  _renderGameScreen() {
    const {questions, maxMistakes, step, onPlayClick, onAnswer, mistakes, onGoToWelcome, userStatus} = this.props;
    const question = questions[step];


    if (step === -1) {
      return (
        <WelcomeScreen
          errorCount={maxMistakes}
          onPlayClick={onPlayClick}
        />
      );
    }

    if (mistakes >= maxMistakes) {
      return <Redirect to={AppRoute.LOSE} />;
    }

    if (step >= questions.length) {
      if (userStatus === AuthorizationStatus.AUTH) {
        return <Redirect to={AppRoute.RESULT} />;
      } else if (userStatus === AuthorizationStatus.NO_AUTH) {
        return <Redirect to={AppRoute.LOGIN} />;
      }
    }

    if (question) {
      switch (question.type) {
        case GameType.ARTIST:
          return (
            <GameScreen gameType={GameType.ARTIST} maxMistakes={maxMistakes} mistakes={mistakes} onGoToWelcome={onGoToWelcome}>
              <QuestionArtistScreenWrapped
                question={question}
                onAnswer={onAnswer}
              />
            </GameScreen>
          );
        case GameType.GENRE:
          return (
            <GameScreen gameType={GameType.GENRE} maxMistakes={maxMistakes} mistakes={mistakes} onGoToWelcome={onGoToWelcome}>
              <QuestionGenreScreenWrapped
                question={question}
                onAnswer={onAnswer}
              />
            </GameScreen>
          );
      }
    }

    return null;
  }

  render() {
    const {questions, onAuthSubmit, onRepeat, mistakes, userStatus} = this.props;

    return (
      <BrowserRouter>
        <Switch>
          <Route exact path={AppRoute.ROOT}>
            {this._renderGameScreen()}
          </Route>
          <Route exact path={AppRoute.LOGIN}>
            <AuthorizationScreen onAuthSubmit={onAuthSubmit} onRepeat={onRepeat} userStatus={userStatus} />
          </Route>
          <Route path={AppRoute.LOSE} exact>
            <GameOverScreen onRepeat={onRepeat} />
          </Route>
          <PrivateRoute exact path={AppRoute.RESULT} render={() => <WinScreen onRepeat={onRepeat} quantity={questions.length} mistakes={mistakes} />}>

          </PrivateRoute>
        </Switch>
      </BrowserRouter>
    );
  }
}

const mapStateToProps = (state) => ({
  mistakes: getMistakes(state),
  maxMistakes: getMaxMistakes(state),
  questions: getQuestions(state),
  step: getStep(state),
  userStatus: getUserStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  onPlayClick: () => {
    dispatch(ActionCreator.incrementStep());
  },
  onAnswer: (question, answer) => {
    dispatch(ActionCreator.incrementMistakes(question, answer));
    dispatch(ActionCreator.incrementStep());
  },
  onRepeat: () => {
    dispatch(ActionCreator.repeatGame());
  },
  onAuthSubmit: (login, password) => {
    dispatch(UserOperation.sendAuthRequest(login, password));
  },
  onGoToWelcome: () => {
    dispatch(ActionCreator.toWelcome());
  }
});

const ConnectedApp = connect(mapStateToProps, mapDispatchToProps)(App);

export {App};
export default ConnectedApp;
