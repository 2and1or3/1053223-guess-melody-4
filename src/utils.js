import {GameType} from './types.ts';

const mockAudioNode = (element) => {
  if (element.type === `audio`) {
    return element;
  }

  return null;
};

const extend = (a, b) => {
  return Object.assign({}, a, b);
};

const adapterForGenreType = (question) => question;

const adapterForArtistType = (question) => ({
  type: question.type,
  artist: question.song.artist,
  trackSrc: question.song.src,
  answers: question.answers.map((answer) => ({
    artist: answer.artist,
    pictureSrc: answer.picture,
  }))
});

const adapterToLocalQuestions = (questions) => {

  const localQuestions = questions.map((question) => {
    switch (question.type) {
      case (GameType.ARTIST):
        return adapterForArtistType(question);
      case (GameType.GENRE):
        return adapterForGenreType(question);

      default:
        return question;
    }
  });

  return localQuestions;
};

export {mockAudioNode, adapterToLocalQuestions, extend};
