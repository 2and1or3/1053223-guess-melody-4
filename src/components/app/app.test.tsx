import * as React from "react";
import * as renderer from "react-test-renderer";

import {App} from './app';
import {mockAudioNode, noop} from '../../utils';
import {GenreQuestion, ArtistQuestion, GameType} from '../../types';


type Question = GenreQuestion | ArtistQuestion;

const questions: Question[] = [
  {
    type: GameType.GENRE,
    genre: `rock`,
    answers: [
      {
        genre: `rock`,
        src: `src-for-track-1`,
      },
      {
        genre: `jazz`,
        src: `src-for-track-2`,
      },
      {
        genre: `pop`,
        src: `src-for-track-3`,
      },
      {
        genre: `alternative`,
        src: `src-for-track-4`,
      }
    ]
  },
  {
    type: GameType.ARTIST,
    artist: `true-person`,
    trackSrc: `src-for-track`,
    answers: [
      {
        artist: `true-person`,
        pictureSrc: `http://placehold.it/134x134`,
      },
      {
        artist: `person-2`,
        pictureSrc: `http://placehold.it/134x134`,
      },
      {
        artist: `person-3`,
        pictureSrc: `http://placehold.it/134x134`,
      },
    ]
  },
];

const commonProps = {
  maxMistakes: 3,
  questions,
  mistakes: 0,
  step: -1,
  onPlayClick: noop,
  onAnswer: noop,
  onRepeat: noop,
  onAuthSubmit: noop,
  userStatus: `NO_AUTH`,
  onGoToWelcome: noop,
};

describe(`Render App component`, () => {
  it(`Render WelcomeScreen component`, () => {
    const tree = renderer
      .create(<App {...commonProps} step={-1} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Render QuestionGenreScreen component`, () => {
    const tree = renderer
      .create(<App {...commonProps} step={0} />, {
        createNodeMock: mockAudioNode
      }
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Render QuestionArtistScreen component`, () => {
    const tree = renderer
      .create(<App {...commonProps} step={1} />, {
        createNodeMock: mockAudioNode
      }
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
