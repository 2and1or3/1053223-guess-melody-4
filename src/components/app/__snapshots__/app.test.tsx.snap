// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render App component Render QuestionArtistScreen component 1`] = `
<section
  className="game game--artist"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="/"
      onClick={[Function]}
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
        style={
          Object {
            "filter": "url(#blur)",
            "transform": "rotate(-90deg) scaleY(-1)",
            "transformOrigin": "center",
          }
        }
      />
    </svg>
    <div
      className="game__mistakes"
    >
      <div
        className=""
      />
      <div
        className=""
      />
      <div
        className=""
      />
    </div>
  </header>
  <section
    className="game__screen"
  >
    <h2
      className="game__title"
    >
      Кто исполняет эту песню?
    </h2>
    <div
      className="game__track"
    >
      <div
        className="track"
      >
        <button
          className="track__button track__button--pause"
          disabled={true}
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
      </div>
    </div>
    <form
      className="game__artist"
    >
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id="true-person0"
          name="answer"
          onChange={[Function]}
          type="radio"
          value="true-person"
        />
        <label
          className="artist__name"
          htmlFor="true-person0"
        >
          <img
            alt="true-person"
            className="artist__picture"
            src="http://placehold.it/134x134"
          />
          true-person
        </label>
      </div>
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id="person-21"
          name="answer"
          onChange={[Function]}
          type="radio"
          value="person-2"
        />
        <label
          className="artist__name"
          htmlFor="person-21"
        >
          <img
            alt="person-2"
            className="artist__picture"
            src="http://placehold.it/134x134"
          />
          person-2
        </label>
      </div>
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id="person-32"
          name="answer"
          onChange={[Function]}
          type="radio"
          value="person-3"
        />
        <label
          className="artist__name"
          htmlFor="person-32"
        >
          <img
            alt="person-3"
            className="artist__picture"
            src="http://placehold.it/134x134"
          />
          person-3
        </label>
      </div>
    </form>
  </section>
</section>
`;

exports[`Render App component Render QuestionGenreScreen component 1`] = `
<section
  className="game game--genre"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="/"
      onClick={[Function]}
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
        style={
          Object {
            "filter": "url(#blur)",
            "transform": "rotate(-90deg) scaleY(-1)",
            "transformOrigin": "center",
          }
        }
      />
    </svg>
    <div
      className="game__mistakes"
    >
      <div
        className=""
      />
      <div
        className=""
      />
      <div
        className=""
      />
    </div>
  </header>
  <section
    className="game__screen"
  >
    <h2
      className="game__title"
    >
      Выберите 
      rock
       треки
    </h2>
    <form
      className="game__tracks"
      onSubmit={[Function]}
    >
      <div
        className="track"
      >
        <button
          className="track__button track__button--pause"
          disabled={true}
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
        <div
          className="game__answer"
        >
          <input
            checked={false}
            className="game__input visually-hidden"
            id="rock0"
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="rock"
          />
          <label
            className="game__check"
            htmlFor="rock0"
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          disabled={true}
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
        <div
          className="game__answer"
        >
          <input
            checked={false}
            className="game__input visually-hidden"
            id="jazz1"
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="jazz"
          />
          <label
            className="game__check"
            htmlFor="jazz1"
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          disabled={true}
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
        <div
          className="game__answer"
        >
          <input
            checked={false}
            className="game__input visually-hidden"
            id="pop2"
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="pop"
          />
          <label
            className="game__check"
            htmlFor="pop2"
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          disabled={true}
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
        <div
          className="game__answer"
        >
          <input
            checked={false}
            className="game__input visually-hidden"
            id="alternative3"
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="alternative"
          />
          <label
            className="game__check"
            htmlFor="alternative3"
          >
            Отметить
          </label>
        </div>
      </div>
      <button
        className="game__submit button"
        type="submit"
      >
        Ответить
      </button>
    </form>
  </section>
</section>
`;

exports[`Render App component Render WelcomeScreen component 1`] = `
<section
  className="welcome"
>
  <div
    className="welcome__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <button
    className="welcome__button"
    onClick={[Function]}
  >
    <span
      className="visually-hidden"
    >
      Начать игру
    </span>
  </button>
  <h2
    className="welcome__rules-title"
  >
    Правила игры
  </h2>
  <p
    className="welcome__text"
  >
    Правила просты:
  </p>
  <ul
    className="welcome__rules-list"
  >
    <li>
      Нужно ответить на все вопросы.
    </li>
    <li>
      Можно допустить 
      3
       ошибки.
    </li>
  </ul>
  <p
    className="welcome__text"
  >
    Удачи!
  </p>
</section>
`;
